go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2628,6 +2631,294 @@ func (s *VtctldServer) ValidateShard(ctx context.Context, req *vtctldatapb.Valid
+			}
+		}
+
+		aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, req.Keyspace, shard)
+		if err != nil {
+			rec.RecordError(fmt.Errorf("FindAllTabletAliasesInShard(%v, %v) failed: %v", req.Keyspace, shard, err))
+			continue
+		}
+
+		for _, alias := range aliases {
+			// Don't diff schemas for self
+			if referenceAlias == alias {
+				continue
+			}
+			wg.Add(1)
+			go func(alias *topodatapb.TabletAlias) {
+				defer wg.Done()
+				if err := s.diffSchema(ctx, referenceSchema, referenceAlias, alias, req.ExcludeTables, req.IncludeViews); err != nil {
+					rec.RecordError(err)
+				}
+			}(alias)
+		}
+	}
+
+	wg.Wait()
+
+	if rec.HasErrors() {
+		return nil, vterrors.Wrapf(rec.Error(), "schema diffs: %v", rec.Error().Error())
+	}
+	return &vtctldatapb.ValidateSchemaKeyspaceResponse{}, nil
+}
+
+// ValidateSchemaShard is part of the vtctlservicepb.VtctldServer interface
+func (s *VtctldServer) ValidateSchemaShard(ctx context.Context, req *vtctldatapb.ValidateSchemaShardRequest) (*vtctldatapb.ValidateSchemaShardResponse, error) {
+	span, ctx := trace.NewSpan(ctx, "VtctldServer.ValidateSchemaShard")
+	defer span.Finish()
+
+	span.Annotate("keyspace", req.Keyspace)
+	span.Annotate("shard", req.Shard)
+	span.Annotate("exclude_tables", strings.Join(req.ExcludeTables, ","))
+	span.Annotate("include_vschema", req.IncludeVschema)
+	span.Annotate("include_views", req.IncludeViews)
+
+	si, err := s.ts.GetShard(ctx, req.Keyspace, req.Shard)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetShard(%v, %v) failed: %v", req.Keyspace, req.Shard, err)
+	}
+
+	if !si.HasPrimary() {
+		return nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no primary in shard %v/%v", req.Keyspace, req.Shard)
+	}
+	log.Infof("Gathering schema for primary %v", topoproto.TabletAliasString(si.Shard.PrimaryAlias))
+
+	getTabletReq := &vtctldatapb.GetTabletRequest{
+		TabletAlias: si.Shard.PrimaryAlias,
+	}
+
+	getTabletResponse, err := s.GetTablet(ctx, getTabletReq)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetTablet(%v) failed: %v", topoproto.TabletAliasString(getTabletReq.TabletAlias), err)
+	}
+
+	primarySchema, err := s.tmc.GetSchema(ctx, getTabletResponse.Tablet, nil, req.ExcludeTables, req.IncludeViews)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetSchema(%v, nil, %v, %v) failed: %v", topoproto.TabletAliasString(si.Shard.PrimaryAlias), req.ExcludeTables, req.IncludeViews, err)
+	}
+
+	if req.IncludeVschema {
+		validateVSchemaReq := &vtctldatapb.ValidateVSchemaRequest{
+			Keyspace:      req.Keyspace,
+			Shards:        []string{req.Shard},
+			ExcludeTables: req.ExcludeTables,
+			IncludeViews:  req.IncludeViews,
+		}
+
+		if _, err := s.ValidateVSchema(ctx, validateVSchemaReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateVSchema(%v) failed: %v", validateVSchemaReq, err)
+		}
+	}
+
+	// read all the aliases in the shard, that is all tablets that are
+	// replicating from the primary
```

Comment:
I know you copied this over, but it's both confusing and inaccurate. Better would be something like:

> Read all the aliases in the shard. Later we are going to filter out the primary, so we don't compare the primary schema with itself.

Technically this call doesn't just get the replicas; it gets everything, and then we manually skip over the primary inside the for loop further down.